Friend endpoints

XuechunHou opened the pull request:
This PR contains :
1. Modification on Author model (the new way of implementing the Author model is that extended built-in User model and added custom fields)
2. Added a new Model "Friend" 
3. http://service/author/<authorid>/friends endpoint implementation
4. http://service/author/<authorid>/friends/<authorid2> endpoint implementation
5. http://service/author/<authorid> endpoint implementation

XuechunHou commented:
This PR also includes:
6. created a new model "FriendRequest"
7. added additional endpoints for internal use:
     http://service/friendrequest/handle: which accepts friend request on POST, rejects friend request on DELTE
    http://service/friendrequest/<str:author_id>: which GET a list of friend requests of current author 
    http://service/author/unfriend: which when POST unfriend current author with required author 
    http://service/author/<str:author_id>/addfriend which GET a list of authors in database which hasn't friended with current author.

AustinGrey commented on web_server/comments/migrations/0002_comment_content.py:
```diff
@@ -1,19 +0,0 @@
-# Generated by Django 3.0.2 on 2020-02-01 04:31
```
Why are you deleting migration files? These need to stay around so that everyone can migrate their existing DB to the schema you have defined.

XuechunHou replied:
I regenerated new migration file for Comment, cuz previously Comment is linking old Author model and I changed it so that it references the new Author model.

AustinGrey replied:
I see, ok then, not really sure how to get around this then -> that means this is a breaking change. If I understand migrations correctly if someone doesn't have the latest set of migrations and then attempts to migrate they might not be able to - will have to keep an eye out after we merge if someone says they get an error when migrating.

AustinGrey commented on web_server/users/models.py:
```diff
@@ -1,12 +1,19 @@
+
+    bio = models.TextField()
+    host = models.URLField()
+    display_name = models.CharField(max_length=256)
```
Why are first and last name not fields in this model anymore?

XuechunHou replied:
if u could please check the built in users_author model 
```
.schema users_author
CREATE TABLE IF NOT EXISTS "users_author" ("password" varchar(128) NOT NULL, "last_login" datetime NULL, "is_superuser" bool NOT NULL, "username" varchar(150) NOT NULL UNIQUE, "first_name" varchar(30) NOT NULL, "last_name" varchar(150) NOT NULL, "email" varchar(254) NOT NULL, "is_staff" bool NOT NULL, "is_active" bool NOT NULL, "date_joined" datetime NOT NULL, "id" char(32) NOT NULL PRIMARY KEY, "uid" varchar(500) NOT NULL, "bio" text NOT NULL, "host" varchar(200) NOT NULL, "display_name" varchar(256) NOT NULL, "url" varchar(500) NOT NULL, "github" varchar(200) NOT NULL);
sqlite>
```

The first_name and last_name column already defined, here I just added additional column that doesn't come with.

AustinGrey commented on web_server/posts/models.py:
```diff
@@ -11,7 +11,8 @@ class Category(models.Model):
 
     def __str__(self):
-        return f'{self.name}'
+        return "{}".format(self.name)
```
Any particular reason you do not want to use f strings? They are succinct, and supported by the version of python we are supposed to target.

XuechunHou replied:
if you require everyone's Python version to be at least 3.6 +, I am fine with using f string.

AustinGrey commented on web_server/posts/models.py:
```diff
@@ -74,16 +76,21 @@ class Post(models.Model):
\ No newline at end of file
+        # number of chars to show in content snippet before cutting off with elipsis
+        post_snippet_length = 15
+        result = "{} post by {}: ".format(self.visibility, self.author)
```
Same issue with f string and formatting here. I notice additional line breaks are being added as well. Are my lines too long?

AustinGrey commented on web_server/friendship/models.py:
```diff
@@ -1,3 +1,23 @@
+        unique_together = (("author_id", "friend_id"),)
+
+
+class FriendRequest(models.Model):
```
FriendRequest wasn't one of the models we discussed during our meeting last week. I'm fine if we add this model, but I'm uncertain how you plan to use it, and your commit lacks documentation regarding it. Please document this model.

XuechunHou replied:
Friend request is the model that is intended to store incoming friend requests. 
`CREATE TABLE IF NOT EXISTS "friendship_friendrequest" ("id" integer NOT NULL PRIMARY KEY AUTOINCREMENT, "from_id" varchar(500) NOT NULL, "to_id" varchar(500) NOT NULL);
CREATE UNIQUE INDEX "friendship_friendrequest_from_id_to_id_c10e9741_uniq" ON "friendship_friendrequest" ("from_id", "to_id");`

XuechunHou replied:
so the friend request is sent from `from_id` to  `to_id`. Both are URLfield.

XuechunHou commented:
1. f string only supported in Python 3.6+, if everyone agreed to use Python 3.6+ , I am fine with using f string. 
2. I regenerated the migration file for Comment model as it was referencing the old Author model and I changed it so that it references the new Author model.  I thought it's better to have one migration file that generates the model completely and accurately instead of having several patches..

3. I added a New Model FriendRequest which is intended to store incoming (and outgoing) friend requests. 
`schema friendship_friendrequest
CREATE TABLE IF NOT EXISTS "friendship_friendrequest" ("id" integer NOT NULL PRIMARY KEY AUTOINCREMENT, "from_id" varchar(500) NOT NULL, "to_id" varchar(500) NOT NULL);
CREATE UNIQUE INDEX "friendship_friendrequest_from_id_to_id_c10e9741_uniq" ON "friendship_friendrequest" ("from_id", "to_id");`

AustinGrey commented:
1.  3.6+ is the python version in the VM, we should use the same python version that is required in the assignments in the class. Would you please revert these changes to f-strings?
2. The point of migrations as I understand it **is** to have the several patches -> it keeps a history record of changes to the database so that if someone checks out the project after having the project version several patches behind it will upgrade their database without data loss. Right now that's not a huge problem but in the future it might leave one group member without a way to migrate and they will need to recreate their database and lose test data.
3. Thank you, I don't think we have a central place to document information like this, so would you please insert your documentation as a docstring on the model itself?

XuechunHou commented:
1. I reversed back to f-strings. 
2. I added documentation to Friend, FriendRequest and Author model
